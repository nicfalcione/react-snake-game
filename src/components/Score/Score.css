.game-score {
    font-size: 2rem;
    color: white;
}