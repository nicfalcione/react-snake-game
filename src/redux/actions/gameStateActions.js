import {
    CHECK_IF_HIT_APPLE,
    CHECK_IF_HIT_SNAKE, GAME_OVER, INCREASE_SPEED, MOVE, NEW_GAME, PAUSE_GAME,
    RESUME_GAME, SET_APPLE_POSITION, SET_DIRECTION
} from '../../constants/actionNames';

export const gameOver = () => ({type: GAME_OVER});
export const pauseGame = () => ({type: PAUSE_GAME});
export const resumeGame = () => ({type: RESUME_GAME});
export const newGame = () => ({type: NEW_GAME});

export const move = (value) => {
    return {
        type: MOVE,
        payload: value
    };
};

export const setDirection = (value) => {
    return {
        type: SET_DIRECTION,
        payload: value
    };
};

export const setApplePosition = (value) => {
    return {
        type: SET_APPLE_POSITION,
        payload: value
    };
};

export const checkIfHitApple = (value) => {
    return {
        type: CHECK_IF_HIT_APPLE,
        payload: value
    };
};

export const checkIfHitSnake = () => ({type: CHECK_IF_HIT_SNAKE});
export const increaseSpeed = () => ({type: INCREASE_SPEED});