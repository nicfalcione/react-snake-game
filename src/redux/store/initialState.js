import { RIGHT } from '../../constants/constants';

const defaultState = {
    isRunning: true,
    gameOver: false,
    direction: RIGHT,
    score: 0,
    speed: 100,
    snakeSegments:[
        {
            key: 1,
            xPos: 380,
            yPos: 220
        },
        {
            key: 2,
            xPos: 360,
            yPos: 220
        },
        {
            key: 3,
            xPos: 340,
            yPos: 220
        }
    ],
    apple: {
        xPos: 400,
        yPos: 400
    }
};

export default defaultState;