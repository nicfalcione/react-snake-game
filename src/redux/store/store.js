import { createStore } from '@reduxjs/toolkit';
import { gameStateReducer } from '../reducers/gameStateReducer';

const store = createStore(gameStateReducer);

export default store;