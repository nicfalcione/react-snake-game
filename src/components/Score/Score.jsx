import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { increaseSpeed, pauseGame, resumeGame } from '../../redux/actions/gameStateActions';
import './Score.css';

const Score = ({ score }) => {
    const dispatch = useDispatch();

    const { isRunning, gameOver } = useSelector((state) => ({
        isRunning: state.isRunning,
        gameOver: state.gameOver
    }));

    const formatScore = (score) => {
        return `Score: ${score}`;
    }

    const onClickScore = (e) => {
        if (!gameOver) {
            if (isRunning) {
                dispatch(pauseGame());
            } else {
                dispatch(resumeGame());
            }
        }
    }

    useEffect(() => {
        if (score > 0 && score % 500 === 0) {
            dispatch(increaseSpeed());
        }
    }, [score]);

    return (
        <span className="game-score" onClick={onClickScore}>{formatScore(score)}</span>
    );
}

export default Score;