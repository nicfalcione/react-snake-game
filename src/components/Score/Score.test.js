import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { gameOver } from "../../redux/actions/gameStateActions";
import store from '../../redux/store/store';
import Score from './Score';

const setupRender = (score) => {
    return render(
        <Provider store={store}><Score score={score}/></Provider>
    )
};

describe('Score component tests', () => {
    it('should show the score formatted on render', () => {
        const { getByText } = setupRender(0);
        expect(getByText('Score: 0')).toBeInTheDocument();
    });

    it('should pause/resume the game when clicking the score', async () => {
        const { getByText } = setupRender(300);
        act(() => {
            fireEvent.click(getByText('Score: 300'));
        });
        expect(store.getState().isRunning).toBeFalsy();

        act(() => {
            fireEvent.click(getByText('Score: 300'));
        }); 
        expect(store.getState().isRunning).toBeTruthy();
    });

    it('should do nothing when clicking the score if the game is over', async () => {
        const { getByText } = setupRender(500);
        act(() => {
            store.dispatch(gameOver());
            fireEvent.click(getByText('Score: 500'));
        });
        expect(store.getState().isRunning).toBeFalsy();
    });

    xit('should increase speed when the score hits 500', () => {
        const { getByText } = setupRender(0);
        expect(store.getState().speed).toBe(100);
    });

    it('should match snapshot', () => {
        const { container } = setupRender();
        expect(container).toMatchSnapshot();
    });
})