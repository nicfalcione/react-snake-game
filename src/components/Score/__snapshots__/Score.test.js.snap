// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score component tests should match snapshot 1`] = `
<div>
  <span
    class="game-score"
  >
    Score: undefined
  </span>
</div>
`;
